src/main/scala/akka/persistence/jdbc/query/scaladsl/JdbcReadJournal.scala

```diff
@@ -135,39 +136,104 @@ class JdbcReadJournal(config: Config)(implicit val system: ExtendedActorSystem)
+    if (latestOrdering.maxOrdering < offset) Source.empty
+    else {
+      readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
+        .mapAsync(1)(Future.fromTry)
+        .mapConcat {
+          case (repr, _, row) =>
+            adaptEvents(repr).map(r => EventEnvelope(Sequence(row.ordering), r.persistenceId, r.sequenceNr, r.payload))
+        }
+    }
+  }
+
+  /**
+   * @param terminateAfterOffset If None, the stream never completes. If a Some, then the stream will complete once a
+   *                             query has been executed which might return an event with this offset (or a higher offset).
+   *                             The stream may include offsets higher than the value in terminateAfterOffset, since the last batch
+   *                             will be returned completely.
+   */
+  private def eventsByTag(tag: String, offset: Long, terminateAfterOffset: Option[Long]): Source[EventEnvelope, NotUsed] = {
+
     import akka.pattern.ask
-    implicit val askTimeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
-    Source.fromFuture(orderingActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId])
-      .flatMapConcat { latestOrdering =>
-        if (latestOrdering.maxOrdering < offset) Source.empty
-        else readJournalDao.eventsByTag(tag, offset, latestOrdering.maxOrdering, max)
-      }
-      .mapAsync(1)(Future.fromTry)
+    implicit val askTimeout: Timeout = Timeout(readJournalConfig.journalSequenceRetrievalConfiguration.askTimeout)
+
+    sealed trait FlowControl
+
+    /** Keep querying - used when we are sure that there is more events to fetch */
+    object Continue extends FlowControl
+
+    /**
+     * Keep querying with delay - used when we have consumed all events,
+     * but want to poll for future events
+     */
+    object ContinueDelayed extends FlowControl
+
+    /** Stop querying - used when we reach the desired offset  */
+    object Stop extends FlowControl
+
+    Source.unfoldAsync[(Long, FlowControl), Seq[EventEnvelope]]((offset, Continue)) {
+
+      case (from, control) =>
+
+        val batchSize = readJournalConfig.maxBufferSize
+
+        def retrieveNextBatch() = {
+          for {
+            queryUntil <- journalSequenceActor.ask(GetMaxOrderingId).mapTo[MaxOrderingId]
+            xs <- currentJournalEventsByTag(tag, from, batchSize, queryUntil).runWith(Sink.seq)
+          } yield {
+
+            // calculates if we have more events to fetch
+            val hasMoreEvents = {
+
+              val possibleRangeSize = queryUntil.maxOrdering - from
+
+              // when batch is being limited by queryUntil and we did fetch everything we could,
+              // we can consider that there is nothing else to fetch
+              if (possibleRangeSize <= batchSize && xs.size == possibleRangeSize) false
+              // otherwise, base decision solely on batch size
+              else xs.size == batchSize
+            }
```

Comment:
The `hasMoreEvents` tried to identify we there are more events to fetch. 

It covers the following cases:

### Scenario 1
Batch was limited by `queryUntil` making it smaller than configured `batchSize`. 
That means that fetched everything we could and we can consider

### Scenario 2
Batch was limited by the batch size and we fetched as much as batch size. 
In that case we can't know if there are more events or not, so we assume there are.

### Scenario 3
Batch was limited by the batch size and we fetched less items.
In that case can consider that we reached the end of the stream.